util/bigint/src/uint.rs

```diff
@@ -425,12 +425,16 @@ macro_rules! uint_overflowing_mul_reg {
 				let (c_u, overflow_u) = mul_u32(a, b_u, c_l >> 32);
 				ret[i + j] = (c_l & 0xFFFFFFFF) + (c_u << 32);
 
-				// Only single overflow possible here
-				let carry = (c_u >> 32) + (overflow_u << 32) + overflow_l + carry2;
-				let (carry, o) = carry.overflowing_add(ret[i + j + 1]);
-
-				ret[i + j + 1] = carry;
-				carry2 = o as u64;
+				// No overflow here
+				let res = (c_u >> 32) + (overflow_u << 32);
+				// possible overflows
+				let (res, o1) = res.overflowing_add(overflow_l);
+				let (res, o2) = res.overflowing_add(carry2);
+				let (res, o3) = res.overflowing_add(ret[i + j + 1]);
+				ret[i + j + 1] = res;
+
+				// Only single overflow possible there
+				carry2 = (o1 | o2 | o3) as u64;
```

Comment:
We are multiplying two 64bit numbers - the result will fit in 128bits.
Those operations are adding the possible carrry (so two 64bits, split into couple of parts) - the result will fit in 64bits + 1 carry.